### PR TITLE
Support for Activity tasks in TaskPoller

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

Added support for activity tasks in new TaskPoller.

Also, added support for providing a custom poll request (instead of various options).

Not Ported:

- activity task cancellation was not ported over; not a single existing test seems to be using it
- handling activity tasks by ID is only used twice across the entire codebase; not ported over for now

## How did you test it?

Two existing tests have been ported over: one with a successful task response and one with a failure.

### CODE REVIEW

File: tests/activity_test.go [hunk 1]
```diff
@@ -646,24 +633,21 @@ func (s *ActivityTestSuite) TestActivityRetry() {
...
 	descResp, err = describeWorkflowExecution()
 	s.NoError(err)
-	for _, pendingActivity := range descResp.GetPendingActivities() {
-		if pendingActivity.GetActivityId() == "A" {
-			s.NotNil(pendingActivity.GetLastFailure().GetApplicationFailureInfo())
-			expectedErrString := "bad-bug"
-			s.Equal(expectedErrString, pendingActivity.GetLastFailure().GetMessage())
-			s.True(pendingActivity.GetLastFailure().GetApplicationFailureInfo().GetNonRetryable())
-			s.Equal(identity2, pendingActivity.GetLastWorkerIdentity())
-		}
```
Comment (stephanos) [inline tests/activity_test.go:665]: Note: this code was never executed before these changes.

File: tests/update_workflow_test.go
Comment (stephanos) [inline tests/update_workflow_test.go:1]: Example for using `WithWorkflowTaskPollRequest`

File: common/testing/taskpoller/taskpoller.go [hunk 2]
```diff
@@ -47,29 +48,47 @@ type (
...
+		return func(o *Options) {
+			o.pollActivityTaskRequest = req
+		}
+	}
+	// WithWorkflowTaskPollRequest defines the PollWorkflowTaskQueueRequest to use for polling
+	WithWorkflowTaskPollRequest = func(req *workflowservice.PollWorkflowTaskQueueRequest) OptionFunc {
+		return func(o *Options) {
+			o.pollworkflowTaskRequest = req
+		}
+	}
```
Comment (stephanos) [inline common/testing/taskpoller/taskpoller.go:84]: I considered not using OptionFunc here and instead make it a require argument, but it seems like this will only be used for special reasons (like worker versioning).

File: common/testing/taskpoller/taskpoller.go [hunk 3]
```diff
@@ -46,30 +47,42 @@ type (
...
 		namespace string
 	}
+	workflowTaskPoller struct {
+		*TaskPoller
+		pollWorkflowTaskRequest *workflowservice.PollWorkflowTaskQueueRequest
+	}
+	activityTaskPoller struct {
+		*TaskPoller
+		pollActivityTaskRequest *workflowservice.PollActivityTaskQueueRequest
+	}
```
Comment (stephanos) [inline common/testing/taskpoller/taskpoller.go:57]: Deliberately un-exported to force DSL-like usage.

File: common/testing/taskpoller/taskpoller.go [hunk 4]
```diff
@@ -46,30 +47,41 @@ type (
...
+	workflowTaskPoller struct {
+		*TaskPoller
+		pollWorkflowTaskRequest *workflowservice.PollWorkflowTaskQueueRequest
 	}
-	// OptionFunc is a function to change an Options instance
-	OptionFunc func(*Options)
+	activityTaskPoller struct {
+		*TaskPoller
+		pollActivityTaskRequest *workflowservice.PollActivityTaskQueueRequest
+	}
```
Comment (stephanos) [inline common/testing/taskpoller/taskpoller.go:57]: Deliberately un-exported to force DSL-like usage.

File: common/testing/taskpoller/taskpoller.go [hunk 5]
```diff
@@ -84,70 +96,139 @@ func New(
...
+// PollAndHandleWorkflowTask issues a PollWorkflowTaskQueueRequest to obtain a new workflow task,
 // invokes the handler with the task, and completes/fails the task accordingly.
+// Any unspecified but required request and response fields are automatically generated using `tv`.
+// Returning an error from `handler` fails the task.
+// If no task is available, it returns NoTaskAvailable.
 func (p *TaskPoller) PollAndHandleWorkflowTask(
 	tv *testvars.TestVars,
 	handler func(task *workflowservice.PollWorkflowTaskQueueResponse) (*workflowservice.RespondWorkflowTaskCompletedRequest, error),
-	funcs ...OptionFunc,
+	funcs ...optionFunc,
```
Comment (alexshtin) [inline common/testing/taskpoller/taskpoller.go:114]: This should be 
```suggestion
	opts ...optionFunc,
```
everywhere.

File: common/testing/taskpoller/taskpoller.go [hunk 6]
```diff
@@ -46,30 +47,41 @@ type (
...
+	WithTimeout = func(timeout time.Duration) optionFunc {
+		return func(o *options) {
 			o.timeout = timeout
 		}
 	}
-	// WithPollSticky will make the poller use the sticky task queue instead of the normal task queue
-	WithPollSticky OptionFunc = func(o *Options) {
-		o.pollStickyTaskQueue = true
-	}
+	NoTaskAvailable = errors.New("no task available")
```
Comment (alexshtin) [inline common/testing/taskpoller/taskpoller.go:84]: I always hated this message :-). Can you:
1. Separate WFT and AT by explicitly saying what task is not available.
2. Prefix error message with "test task poller:..." or something like this to be explicitly clear that this message is coming from tests.
3. Maybe rephrase entire message to be more explicit: "PollActivityTask API didn't return any task after waiting for X seconds. It means that ActivityTask was never created." I am not sure about exact text, but I want this message to be more actionable for someone who is not familiar with the system.

File: common/testing/taskpoller/taskpoller.go [hunk 7]
```diff
@@ -256,13 +337,113 @@ func (p *TaskPoller) respondWorkflowTaskFailed(
...
+
+func (p *activityTaskPoller) pollAndHandleTask(
+	ctx context.Context,
+	opts *options,
+	handler func(task *workflowservice.PollActivityTaskQueueResponse) (*workflowservice.RespondActivityTaskCompletedRequest, error),
+) (*workflowservice.RespondActivityTaskCompletedResponse, error) {
+	p.t.Helper()
+	task, err := p.pollActivityTask(ctx, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to poll workflow task: %w", err)
```
Comment (alexshtin) [inline common/testing/taskpoller/taskpoller.go:375]:
```suggestion
		return nil, fmt.Errorf("failed to poll activity task: %w", err)
```

File: common/testing/taskpoller/taskpoller.go [hunk 8]
```diff
@@ -84,70 +96,139 @@ func New(
...
-	funcs ...OptionFunc,
-) (*workflowservice.PollWorkflowTaskQueueResponse, error) {
-	p.t.Helper()
-	options := newOptions(tv, funcs)
-	ctx, cancel := newContext(options)
-	defer cancel()
-	return p.pollWorkflowTask(ctx, options)
+	req *workflowservice.PollWorkflowTaskQueueRequest,
+) *workflowTaskPoller {
+	return &workflowTaskPoller{TaskPoller: p, pollWorkflowTaskRequest: req}
```
Comment (alexshtin) [inline common/testing/taskpoller/taskpoller.go:103]: I don't understand this function.

File: common/testing/taskpoller/taskpoller.go [hunk 9]
```diff
@@ -84,70 +96,139 @@ func New(
...
+		req.TaskQueue = opts.tv.TaskQueue()
+	}
+	if req.Identity == "" {
+		req.Identity = opts.tv.WorkerIdentity()
+	}
+	resp, err := p.client.PollWorkflowTaskQueue(ctx, req)
 	if err != nil {
 		return nil, err
 	}
 	if resp == nil || len(resp.TaskToken) == 0 {
```
Comment (alexshtin) [inline common/testing/taskpoller/taskpoller.go:230]: This is probably some legacy but I don't know when `TaskToken` can be 0 length. Let's remove this condition and see if it breaks anything.

**Review by alexshtin — APPROVED** [review]

File: common/testing/taskpoller/taskpoller.go (hunk 8 above)
Comment (stephanos) [reply common/testing/taskpoller/taskpoller.go:104]: The name might need to change. Maybe `WithWorkflowTask` now that we've gotten rid of almost all options.

The idea is that it (a) creates a workflow-specific poller that (b) will use a specific poll request, provided by the caller (instead of the default). There's an example of what that looks like in the PR.

Comment (stephanos) [reply common/testing/taskpoller/taskpoller.go:104]: We decided to leave things as-is and make improvement as we go.
